### PR TITLE
feat(orchestrator): add rate-limited logging for retry attempts (#3109)

### PR DESCRIPTION
**Pull request by RC918**
# feat(orchestrator): add rate-limited logging for retry attempts (#3109)

## Summary

Implements rate-limited logging for transient error retries in `ResilientPostgresSaver` to reduce log noise during prolonged database outages. During extended outages, each retry generates a warning log entry which can overwhelm log systems and make root cause analysis harder.

**Key changes:**
- Add `CHECKPOINT_RETRY_LOG_SAMPLE_RATE` configuration (default: 1 = log all, set to N to log every Nth retry, max: 1000)
- First and last retry attempts are always logged regardless of sample rate
- Add `_total_retry_attempts` counter for metrics (always incremented, even when log is sampled out)
- Add `log_sampled` and `total_retry_attempts` fields to log extra for observability
- `_should_log_retry()` is a pure function with no side effects (counter increments handled by caller)

**Blueprint alignment:** Flow Controller v3 (Fail-Fast Recovery), Telemetry v2 (Observable degradation events)

## Updates since last revision

Addressed code review feedback from Gemini Code Assist and MorningAI Reviewer:

1. **Pure function refactor**: Moved `_retry_log_count` increment out of `_should_log_retry()` into `_retry_with_backoff()` to eliminate side effects
2. **Settings access fix**: Use `getattr()` with default to handle mocked settings in tests
3. **Increased upper bound**: Changed `le=100` to `le=1000` for extreme outage scenarios
4. **Updated test**: Modified `test_should_log_retry_method_directly` to use new `_should_log_retry(is_first, is_last)` signature
5. **Defensive type check**: Added `try/except` around `int(raw_sample_rate)` to handle MagicMock (which returns MagicMock for any attribute access, not the default value from getattr). Falls back to default on TypeError/ValueError.

## Review & Testing Checklist for Human

- [ ] **Verify off-by-one logic**: Check that `is_last_retry = (attempt == self._max_retries - 1)` correctly identifies the last retry *warning* before exhaustion. Note: This is inside `if attempt < self._max_retries:` block, so it's the last retry that will actually retry (not the final exhausted failure, which is logged separately in the else branch)
- [ ] **Counter accumulation behavior**: The `_total_retry_attempts` and `_retry_log_count` counters are instance-level and accumulate across multiple checkpoint operations. Verify this is acceptable for your metrics use case
- [ ] **Sampling correctness**: With the refactored pure function, verify that `_retry_log_count % sample_rate == 0` produces expected sampling behavior (first=always, last=always, intermediate=every Nth)

**Recommended test plan:**
1. Deploy to staging with `CHECKPOINT_RETRY_LOG_SAMPLE_RATE=5`
2. Simulate database outage (e.g., block Postgres port)
3. Verify logs show first retry, every 5th intermediate retry, and last retry
4. Check that `total_retry_attempts` in log extra reflects actual retry count

### Notes

- Tests could not be run locally due to missing `github` module dependency - relying on CI for verification
- Default behavior (sample_rate=1) maintains backward compatibility by logging all retries
- 11 unit tests added covering rate limiting, counter behavior, and edge cases

---
**Link to Devin run:** https://app.devin.ai/sessions/4609907defbd48af8dd21f9b27b8dbb0
**Requested by:** Ryan Chen (ryan2939z@gmail.com) / @RC918

### CODE REVIEW

File: handoff/20250928/40_App/orchestrator/langgraph_orchestrator.py [hunk 1]
```diff
@@ -766,6 +784,38 @@ def _is_connection_lost_error(self, error_str: str) -> bool:
...
+        the first and last attempts for visibility.
+
+        Args:
+            attempt: Current attempt number (0-indexed)
+            is_last_attempt: True if this is the final retry attempt
+
+        Returns:
+            True if this retry should be logged, False to skip logging
+        """
+        self._retry_log_count += 1
```
Comment (RC918) [inline handoff/20250928/40_App/orchestrator/langgraph_orchestrator.py]: IMPACT: This causes incorrect sampling behavior because `_retry_log_count` is incremented even when the first retry (which always logs) occurs, shifting the modulo cycle incorrectly for future retries. For example, with `retry_log_sample_rate=5`, the next logged retry would be at count=5 (already offset by 1), not every 5th retry as expected. This undermines the reliability of log sampling. SUGGESTED FIX: Only increment `_retry_log_count` when the attempt is not the first one (i.e., move increment inside a condition after checking `attempt == 0`).

File: common/config/settings.py [hunk 2]
```diff
@@ -987,6 +987,17 @@ def orchestrator_jwt_secret(self) -> Optional[str]:
...
+        default=1,
+        ge=1,
+        le=100,
+        alias="CHECKPOINT_RETRY_LOG_SAMPLE_RATE",
+        description="Sample rate for retry warning logs during transient errors. "
+                    "1 = log every retry (default), 10 = log every 10th retry. "
+                    "First and last retries are always logged regardless of this setting. "
+                    "Higher values reduce log noise during prolonged outages. (Issue #3109)"
+    )
+
```
Comment (RC918) [inline common/config/settings.py]: IMPACT: Introduces a new configuration field without a corresponding version bump or backward compatibility guarantee in documentation. If this setting is used in automation, dashboards, or alerting rules, it could lead to misconfiguration. SUGGESTED FIX: Add a comment indicating when this field was introduced (e.g., 'Added in v2025.9') and ensure runbooks/alerts are updated to reflect that missing this config means full logging (safe default).

File: handoff/20250928/40_App/orchestrator/langgraph_orchestrator.py [hunk 3]
```diff
@@ -918,22 +968,32 @@ def _retry_with_backoff(self, operation_name: str, operation: Callable):
...
-                            "attempt": attempt + 1,
-                            "max_retries": self._max_retries + 1,
-                            "delay_seconds": delay,
-                            "is_connection_lost": is_connection_lost,
-                            "masking_failed": masking_failed,
-                        }
-                    )
+                    # Issue #3109: Track total retry attempts for metrics
+                    # This counter is always incremented, even when log is sampled out
+                    self._total_retry_attempts += 1
```
Comment (RC918) [inline handoff/20250928/40_App/orchestrator/langgraph_orchestrator.py]: IMPACT: The counter is instance-scoped, so it only tracks retries within a single operation lifetime. In high-throughput scenarios, this limits observability for system-wide retry trends. SUGGESTED FIX: Consider exposing this metric via a global registry (e.g., Prometheus counter) so it can be aggregated across instances and used for monitoring overall system health.

**Review by RC918 — COMMENTED** [review]
## MorningAI Code Review
